languages/ruby/lib/oso/polar/polar.rb

```diff
@@ -28,11 +31,25 @@ def clear
       # @param name [String]
       # @raise [PolarFileExtensionError] if provided filename has invalid extension.
       # @raise [PolarFileNotFoundError] if provided filename does not exist.
-      def load_file(name)
+      def load_file(name) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
         raise PolarFileExtensionError unless ['.pol', '.polar'].include? File.extname(name)
         raise PolarFileNotFoundError, name unless File.file?(name)
 
-        load_queue << name
+        hash = hash_file(name)
+
+        if @loaded_names.key?(name)
+          raise PolarRuntimeError, "File #{name} has already been loaded." unless @loaded_names[name] != hash
+
+          raise PolarRuntimeError, "A file with the name #{name}, but different contents,
+            has already been loaded."
+        elsif @loaded_contents.key?(hash)
+          raise PolarRuntimeError, "A file with the same contents as #{name} named #{@loaded_contents[hash]}
+            has already been loaded."
+        else
+          File.open(name) { |file| load_str(file.read, filename: name) }
```

Comment:
Could we read the file once, then hash it, and then reuse the existing contents here instead of reading it twice?